forc-pkg/src/source/reg/mod.rs

```diff
@@ -0,0 +1,523 @@
+                })?;
+            }
+        }
+        let path = {
+            let _guard = lock.read()?;
+            manifest::find_within(path, ctx.name())
+                .ok_or_else(|| anyhow!("failed to find package `{}` in {}", ctx.name(), self))?
+        };
+        PackageManifestFile::from_file(path)
+    }
+}
+
+impl source::DepPath for Pinned {
+    fn dep_path(&self, _name: &str) -> anyhow::Result<source::DependencyPath> {
+        bail!("dep_path: registry dependencies are not yet supported");
+    }
+}
+
+impl From<Pinned> for source::Pinned {
+    fn from(p: Pinned) -> Self {
+        Self::Registry(p)
+    }
+}
+
+async fn fetch(fetch_id: u64, pinned: &Pinned, ipfs_node: &IPFSNode) -> anyhow::Result<PathBuf> {
+    let path = with_tmp_fetch_index(
+        fetch_id,
+        &pinned.source.name,
+        &pinned.source,
+        |index_file| async move {
+            let path = registry_package_dir(
+                &pinned.source.namespace,
+                &pinned.source.name,
+                &pinned.source.version,
+            );
+            if path.exists() {
+                let _ = fs::remove_dir_all(&path);
+            }
+            fs::create_dir_all(&path)?;
+
+            let package_entry = index_file.get(&pinned.source.version).ok_or_else(|| {
+                anyhow!(
+                    "Version {} not found for {}",
+                    pinned.source.version,
+                    pinned.source.name
+                )
+            })?;
+
+            let cid = Cid::from_str(&package_entry.source_cid).with_context(|| {
+                format!(
+                    "Invalid CID {}v{}: `{}`",
+                    package_entry.name, package_entry.version, package_entry.source_cid
+                )
+            })?;
+            match ipfs_node {
+                IPFSNode::Local => {
+                    println_action_green("Fetching", "with local IPFS node");
+                    cid.fetch_with_client(&ipfs_client(), &path).await?;
+                }
+                IPFSNode::WithUrl(gateway_url) => {
+                    println_action_green(
+                        "Fetching",
+                        &format!("from {}. Note: This can take several minutes.", gateway_url),
+                    );
+                    cid.fetch_with_gateway_url(gateway_url, &path).await?;
+                }
+            }
+
+            Ok(path)
+        },
+    )
+    .await?;
+    Ok(path)
+}
+
+async fn with_tmp_fetch_index<F, O, Fut>(
+    fetch_id: u64,
+    pkg_name: &str,
+    source: &Source,
+    f: F,
+) -> anyhow::Result<O>
+where
+    F: FnOnce(IndexFile) -> Fut,
+    Fut: std::future::Future<Output = anyhow::Result<O>>,
+{
+    let tmp_dir = tmp_registry_package_dir(fetch_id, pkg_name, &source.version, &source.namespace);
+    if tmp_dir.exists() {
+        let _ = std::fs::remove_dir_all(&tmp_dir);
+    }
+
+    // TODO: remove the clone
```

Comment:
needs to be addressed?

Reply:
I was thinking to but upon looking at this in more detail, I think i will leave it as it is because it is a single string clone and to remove it we need to change the Source declaration to have a lifetime param and that will effect all the possible source types whcih will be a fairly large change to remove single string clone. We can do that if we ever end up without anything to do :)